modules/calib3d/src/rhorefc.cpp

```diff
@@ -0,0 +1,2426 @@
+       pkdPts[2].y == pkdPts[3].y || pkdPts[0].y == pkdPts[2].y ||
+       pkdPts[1].y == pkdPts[3].y || pkdPts[0].y == pkdPts[3].y){
+        return 1;
+    }
+
+    /* If the matches do not satisfy the strong geometric constraint, abort. */
+    /* (0 x 1) * 2 */
+    float cross0s0 = pkdPts[0].y-pkdPts[1].y;
+    float cross0s1 = pkdPts[1].x-pkdPts[0].x;
+    float cross0s2 = pkdPts[0].x*pkdPts[1].y-pkdPts[0].y*pkdPts[1].x;
+    float dots0    = cross0s0*pkdPts[2].x + cross0s1*pkdPts[2].y + cross0s2;
+    float cross0d0 = pkdPts[4].y-pkdPts[5].y;
+    float cross0d1 = pkdPts[5].x-pkdPts[4].x;
+    float cross0d2 = pkdPts[4].x*pkdPts[5].y-pkdPts[4].y*pkdPts[5].x;
+    float dotd0    = cross0d0*pkdPts[6].x + cross0d1*pkdPts[6].y + cross0d2;
+    if(((int)dots0^(int)dotd0) < 0){
+        return 1;
+    }
+    /* (0 x 1) * 3 */
+    float cross1s0 = cross0s0;
+    float cross1s1 = cross0s1;
+    float cross1s2 = cross0s2;
+    float dots1    = cross1s0*pkdPts[3].x + cross1s1*pkdPts[3].y + cross1s2;
+    float cross1d0 = cross0d0;
+    float cross1d1 = cross0d1;
+    float cross1d2 = cross0d2;
+    float dotd1    = cross1d0*pkdPts[7].x + cross1d1*pkdPts[7].y + cross1d2;
+    if(((int)dots1^(int)dotd1) < 0){
+        return 1;
+    }
+    /* (2 x 3) * 0 */
+    float cross2s0 = pkdPts[2].y-pkdPts[3].y;
+    float cross2s1 = pkdPts[3].x-pkdPts[2].x;
+    float cross2s2 = pkdPts[2].x*pkdPts[3].y-pkdPts[2].y*pkdPts[3].x;
+    float dots2    = cross2s0*pkdPts[0].x + cross2s1*pkdPts[0].y + cross2s2;
+    float cross2d0 = pkdPts[6].y-pkdPts[7].y;
+    float cross2d1 = pkdPts[7].x-pkdPts[6].x;
+    float cross2d2 = pkdPts[6].x*pkdPts[7].y-pkdPts[6].y*pkdPts[7].x;
+    float dotd2    = cross2d0*pkdPts[4].x + cross2d1*pkdPts[4].y + cross2d2;
+    if(((int)dots2^(int)dotd2) < 0){
+        return 1;
+    }
+    /* (2 x 3) * 1 */
+    float cross3s0 = cross2s0;
+    float cross3s1 = cross2s1;
+    float cross3s2 = cross2s2;
+    float dots3    = cross3s0*pkdPts[1].x + cross3s1*pkdPts[1].y + cross3s2;
+    float cross3d0 = cross2d0;
+    float cross3d1 = cross2d1;
+    float cross3d2 = cross2d2;
+    float dotd3    = cross3d0*pkdPts[5].x + cross3d1*pkdPts[5].y + cross3d2;
+    if(((int)dots3^(int)dotd3) < 0){
+        return 1;
+    }
+
+    /* Otherwise, accept */
+    return 0;
+}
+
+/**
+ * Compute homography of matches in gathered, packed sample and output the
+ * current homography.
+ */
+
+inline void   RHO_HEST_REFC::generateModel(void){
+    hFuncRefC(curr.pkdPts, curr.H);
+}
+
+/**
+ * Checks whether the model is itself degenerate.
+ * - One test: All elements of the homography are added, and if the result is
+ *   NaN the homography is rejected.
+ */
+
+inline int    RHO_HEST_REFC::isModelDegenerate(void){
+    int degenerate;
+    float* H = curr.H;
+    float f=H[0]+H[1]+H[2]+H[3]+H[4]+H[5]+H[6]+H[7];
+
+    /* degenerate = isnan(f); */
+    degenerate = f!=f;/* Only NaN is not equal to itself. */
```

Comment:
@vpisarev I will make the change that you request here, but I object to it being called "very unsafe"; It is in fact the safest alternative short of using the C/C++ Standard's `isnan()`. The IEEE-754 standard requires this behaviour, so unless the compiler is noncompliant in that respect, it's 100% safe, besides being more portable than the implementation of `cvIsNaN()` which I reviewed. A member of the IEEE-754 committee had [this to say](http://stackoverflow.com/a/1573715/2809095) on the topic on StackOverflow.